src/main/java/net/objecthunter/exp4j/tokenizer/Tokenizer.java

```diff
@@ -133,6 +148,16 @@ private boolean isCloseParentheses(char ch) {
         return ch == ')' || ch == '}' || ch == ']';
     }
 
+    private Collection<String> undefinedVariables = null;
+    
+    public void allowUndefinedVariables(boolean allow) {
+    	this.undefinedVariables = (allow ? new ArrayList<String>() : null);
```

Comment:
ArrayList's add can be O(N), if it needs to reallocate the underlying array. Anyway, it's reasonable to expect the set of unknown variables to be small, so I don't think it really matters.